content/courses/advanced-ada/chapters/exceptions.rst

```diff
@@ -235,6 +235,589 @@ the call to :ada:`Assert` is not ignored.
+so we cannot assign :ada:`Arr_2` (:ada:`From` parameter) to :ada:`Arr_1`
+(:ada:`To` parameter) in the :ada:`Assign` procedure.
+
+Overflow Check
+~~~~~~~~~~~~~~
+
+Operations on scalar objects might lead to overflow, which, if not checked,
+lead to wrong information being computed and stored. Therefore, an overflow
+check verifies that the value of a scalar object is within the base range of
+its type. For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Exceptions.Overflow_Check
+    :class: ada-run-expect-failure
+
+    procedure Show_Overflow_Check is
+       A, B : Integer;
+    begin
+       A := Integer'Last;
+       B := 1;
+
+       A := A + B;
+    end Show_Overflow_Check;
+
+In this example, :ada:`A` already has the last possible value of the
+:ada:`Integer'Base` range, so increasing it by one causes an overflow error.
+
+
+Range Check
+~~~~~~~~~~~
+
+The range check verifies that a scalar value is within a specific range |mdash|
+for instance, the range of a subtype. Let's see an example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Exceptions.Range_Check
+    :class: ada-run-expect-failure
+
+    procedure Show_Range_Check is
+
+       type Integer_Array is array (Positive range <>)
+         of Integer;
+
+       function Value_Of (A : Integer_Array;
+                          I : Integer) return Integer is
+          (A (I));
+
+       Arr : Integer_Array (1 .. 10) := (others => 1);
+       I   : Integer;
+
+    begin
+       I := Arr (11);
+    end Show_Range_Check;
+
+In this example, the range of the input array :ada:`Arr` |mdash| passed to
+parameter :ada:`A` of the :ada:`Value_Of` function |mdash| is 1 to 10. In the
+call to :ada:`Value_Of`, we're trying to access position 11, which is outside
+this range. Therefore, the range check fails.
+
+
+Tag Check
+~~~~~~~~~
+
+The tag check ensures that the tag of a tagged object matches the expected tag
+in a dispatching operation. For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Exceptions.Tag_Check
+    :class: ada-run-expect-failure
+
+    package P is
+
+       type T is tagged null record;
+       type T_Class is access all T'Class;
+
+       function Copy (Obj : T_Class)
+                      return T_Class is
+         (Obj);
+
+       type T1 is new T with null record;
+       type T1_Class is access all T1'Class;
+
+       type T2 is new T with null record;
+       type T2_Class is access all T2'Class;
+
+    end P;
+
+    with P; use P;
+
+    procedure Show_Tag_Check is
+
+       A1 : T1_Class := new T1;
+       A2 : T2_Class := new T2;
+       A3 : T_Class;
+
+    begin
+       A3 := T_Class (A1);
+       A2 := T2_Class (Copy (A3));
+
+    end Show_Tag_Check;
+
+Here, the tag of the object returned by the :ada:`Copy` function is
+:ada:`T1'Tag`. However, we cannot convert an object with this tag to an
+object of :ada:`T2_Class`, so the tag check fails.
+
+
+Accessibility Check
```

Comment:
IMHO this one is too complex for an intro course.

Reply:
Same here: this is an advanced course, so I guess it should be OK.